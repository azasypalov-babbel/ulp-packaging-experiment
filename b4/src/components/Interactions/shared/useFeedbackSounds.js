import { useCallback, useContext, useEffect, useReducer, useRef } from 'react';
import { ServiceContext } from '@src/components/shared/withServices';
import incorrectFeedbackSound from '@assets/sounds/babbel_wrong.mp3';
import correctFeedbackSound from '@assets/sounds/babbel_correct.mp3';
import { useSoundPlayer } from '@src/components/Trainer/shared/useSoundPlayer';
import { checks } from '@src/services/soundService';

const useFeedbackSounds = ({ skipOnActiveSound = true, resetDependencies = [] } = {}) => {
  const { settingsService, soundService } = useContext(ServiceContext);
  const correctSoundInstance = useSoundPlayer(correctFeedbackSound);
  const incorrectSoundInstance = useSoundPlayer(incorrectFeedbackSound);
  const [, forceUpdate] = useReducer((s) => !s, true);
  const skippedFeedbackRef = useRef(false);

  useEffect(() => {
    skippedFeedbackRef.current = false;
    correctSoundInstance.reset();
    incorrectSoundInstance.reset();
  }, [correctSoundInstance, incorrectSoundInstance, ...resetDependencies]);

  const hasPlayed = checks.hasPlayed(correctSoundInstance.getState())
    || checks.hasPlayed(incorrectSoundInstance.getState());

  return [
    useCallback((solved) => {
      const isPlayingAnySound = soundService.getPlayingInstances().length > 0;
      if ((skipOnActiveSound && isPlayingAnySound) || !settingsService.isFeedbackSoundEnabled) {
        skippedFeedbackRef.current = true;
        forceUpdate();
        return;
      }
      const instance = solved ? correctSoundInstance : incorrectSoundInstance;
      instance.play();
    }, [skipOnActiveSound, soundService, settingsService, correctSoundInstance, incorrectSoundInstance]),
    hasPlayed || skippedFeedbackRef.current
  ];
};

export default useFeedbackSounds;
