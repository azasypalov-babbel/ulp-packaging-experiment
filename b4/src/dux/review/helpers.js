import { parseQueryParameters } from '../../lib/urlHelpers';
import { isAnalyzerLibRecognizerSupported } from '../../lib/speechHelper';

const REVIEW_SEARCH = 'REVIEW_SEARCH';
const REVIEW_DUE = 'REVIEW_DUE';

const getSearchQueryParams = () => {
  const queryParameters = parseQueryParameters();

  const mapping = {
    packageId: 'package_id',
    level: 'knowledge_level',
    search: 'q'
  };

  return Object.keys(queryParameters).reduce((acc, item) => {
    if (mapping[item]) {
      acc[mapping[item]] = queryParameters[item];
    }
    return acc;
  }, {});
};

export const getReviewQueryParams = (reviewType) => {
  if (reviewType === REVIEW_DUE) {
    return { filter: 'due' };
  } else if (reviewType === REVIEW_SEARCH) {
    return {
      filter: 'search',
      ...getSearchQueryParams()
    };
  }
};

export const filterInteractionTypes = (types) => {
  const interactionsCountMap = types.reduce((acc, type) => ({
    ...acc,
    [type.id]: type.count
  }), {});

  const interactions = [];

  if (interactionsCountMap['flashcard'] > 0) {
    interactions.push({
      id: 'flashcard',
      count: interactionsCountMap['flashcard']
    });
  }

  if (interactionsCountMap['listen'] > 3) {
    interactions.push({
      id: 'listen',
      count: interactionsCountMap['listen']
    });
  }

  if (isAnalyzerLibRecognizerSupported() && interactionsCountMap['speak'] > 3) {
    interactions.push({
      id: 'speak',
      count: interactionsCountMap['speak']
    });
  }

  if (interactionsCountMap['write'] > 0) {
    interactions.push({
      id: 'write',
      count: interactionsCountMap['write']
    });
  }

  return interactions;
};

export const countReviewItems = (trainers) => {
  return trainers.reduce((trainerTotal, trainer) => {
    return trainerTotal + trainer.item_groups.reduce((itemGroupTotal, itemGroup) => {
      return itemGroupTotal + itemGroup.items.length;
    }, 0);
  }, 0);
};
