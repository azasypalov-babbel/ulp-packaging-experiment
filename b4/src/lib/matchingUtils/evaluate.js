
import levenshteinDistance from './levenshteinDistance';

const averageDistance = (stringA, stringB, distance) =>
  distance / Math.max(stringA.length, stringB.length);

export const getSubstitutionCost = (characterSubstitutionCosts = []) => (a, b) => {
  if (a === b) return 0;

  const substitution = characterSubstitutionCosts
    .find(({ character, replacement }) => a === character && b === replacement);
  if (substitution) {
    return substitution.cost;
  }

  return 1;
};

export const RESULTS = {
  CORRECT: 'CORRECT',
  INCORRECT: 'INCORRECT',
  TYPO: 'TYPO'
};

const TYPO_THRESHOLD = 0.25;

export const evaluate = ({ attemptText, targetText, substitutionCosts = [] }) => {
  if (attemptText === targetText) {
    return RESULTS.CORRECT;
  }

  const targetTokens = targetText.split(/\s+/);
  const attemptTokens = attemptText.split(/\s+/);
  if (targetTokens.length != attemptTokens.length) {
    return RESULTS.INCORRECT;
  }
  let hasTypo = false;
  for (let i = 0; i < targetTokens.length; i += 1) {
    const distance = levenshteinDistance(
      targetTokens[i],
      attemptTokens[i],
      getSubstitutionCost(substitutionCosts)
    ).distance;
    if (averageDistance(targetTokens[i], attemptTokens[i], distance) > TYPO_THRESHOLD) {
      return RESULTS.INCORRECT;
    } else if (distance > 0) {
      hasTypo = true;
    }
  }
  return hasTypo ? RESULTS.TYPO : RESULTS.CORRECT;
};
