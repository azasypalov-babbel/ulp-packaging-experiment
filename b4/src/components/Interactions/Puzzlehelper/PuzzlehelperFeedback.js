import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';

import pickRandom from '../../../lib/pickRandom';
import { RESULTS } from '../../../lib/matchingUtils/evaluate';

import { toggleTransliterationVisibility } from '../../../dux/session/actions';

import Delayed from '../../shared/Delayed';
import FeedbackSheet from '../../shared/FeedbackSheet/FeedbackSheet';
import withTranslations from '../../shared/withTranslations';

import { attemptType } from '../shared/scene/scenePropTypes';

const PuzzlehelperFeedback = (props) => {
  const {
    attempt,
    onProceed,
    onRetry,
    translations
  } = props;

  const { solved, feedbackType } = attempt;

  const getFeedbackMessage = (type) => {
    return {
      [RESULTS.CORRECT]: undefined,
      [RESULTS.INCORRECT]: translations.negativeFeedback,
      [RESULTS.TYPO]: translations.typoFeedback
    }[type];
  };

  if (feedbackType === RESULTS.CORRECT) {
    return <Delayed delay={1000} onFinish={onProceed} />;
  } else {
    return (
      <FeedbackSheet
        data-gap-feedback={true}
        data-solved={solved}
        data-mistaken={!solved}
        feedbackMessage={getFeedbackMessage(feedbackType)}
        onContinueClick={onProceed}
        onTryAgainClick={!solved ? onRetry : undefined}
      />
    );
  }
};

PuzzlehelperFeedback.propTypes = {
  attempt: attemptType,
  onProceed: PropTypes.func.isRequired,
  translations: PropTypes.shape({
    negativeFeedback: PropTypes.string.isRequired,
    typoFeedback: PropTypes.string.isRequired
  }).isRequired,
  inputCommands: PropTypes.shape({
    insert: PropTypes.func.isRequired,
    delete: PropTypes.func.isRequired,
    deleteAll: PropTypes.func.isRequired,
    return: PropTypes.func.isRequired
  }),
  onRetry: PropTypes.func.isRequired
};

const mapDispatchToProps = {
  toggleTransliterationVisibility
};

const getTranslations = (translate, props, instanceId) => {
  const negativeFeedbackMessages = [
    'listening_trainer.feedback.negative.1',
    'listening_trainer.feedback.negative.2',
    'listening_trainer.feedback.negative.3',
    'listening_trainer.feedback.negative.4',
    'listening_trainer.feedback.negative.5'
  ].map(translate);

  return {
    negativeFeedback: pickRandom(negativeFeedbackMessages, instanceId),
    typoFeedback: translate('writing.feedback.typos')
  };
};

export default compose(
  connect(null, mapDispatchToProps),
  withTranslations(getTranslations)
)(PuzzlehelperFeedback);
