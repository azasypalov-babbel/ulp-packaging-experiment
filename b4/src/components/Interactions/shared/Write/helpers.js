import { assesUserInput } from '../../../../lib/matchingUtils/assesUserInput';
import { RESULTS } from '../../../../lib/matchingUtils/evaluate';
import getBestMatch from '../../../../lib/matchingUtils/getBestMatch';

const stripHtml = (s) => s ? s.replace(/(<([^>]+)>)/ig, '') : '';

export const splitBefore = (arr, fn = (char) => char === ' ') =>
  arr.reduce(
    (acc, val, i, arr) => {
      fn(val, i, arr)
        ? acc.push([val])
        : acc[acc.length - 1].push(val);
      return acc;
    }, [[]]
  );

export const insertText = (text) => {
  /**
   * This is a custom function to implement what insertText does natively.
   *
   * document.execCommand('insertText', false, text)
   *
   * The command does not work for Firefox because of a bug.
   * https://bugzilla.mozilla.org/show_bug.cgi?id=1220696
   *
   * When that is fixed we can return to using insertText execCommand.
   */
  const activeElement = document.activeElement;
  if (activeElement.tagName === 'INPUT') {
    const start = activeElement.selectionStart;
    const end = activeElement.selectionEnd;
    const value = activeElement.value;

    activeElement.value = activeElement.value.slice(0, start) +
      text + activeElement.value.slice(end);
    const newEnd = start + text.length;

    activeElement.selectionStart = newEnd;
    activeElement.selectionEnd = newEnd;

    /**
     * Manually fire change event
     * tracker.setValue() is to ensure React recognises the change event.
     *
     * https://github.com/facebook/react/issues/10135#issuecomment-500929024
     */
    // eslint-disable-next-line no-underscore-dangle
    const tracker = activeElement._valueTracker;
    if (tracker) {
      tracker.setValue(value);
    }

    let event;
    if (typeof Event === 'function') {
      event = new Event('change', { bubbles: true });
    } else {
      event = document.createEvent('Event');
      event.initEvent('change', true, true);
    }

    activeElement.dispatchEvent(event);
  }
};

export const getAlternativeSolution = (allTexts, firstSolution) => {
  var alternativeTexts = [];
  if (allTexts.length >= 2) {
    alternativeTexts = allTexts.filter(function(text) {
      if (text !== firstSolution) {
        return text;
      }
    });
  }
  if (alternativeTexts.length === 0) {
    return null;
  }
  return stripHtml(alternativeTexts[0]);
};

export const getClosestMatch = ({ text, targetTexts }) => {
  if (!text) {
    // return the first target text for empty attempt text
    return {
      transformedText: '',
      targetText: targetTexts[0],
      solvedPrefix: ''
    };
  }

  const sanitizedText = text.trim();
  const targetText = getBestMatch(sanitizedText, targetTexts).text;
  const solvedPrefix = sanitizedText.split('').reduce((reduced) => {
    if (targetText.toLowerCase().indexOf(reduced.toLowerCase()) === 0) {
      return reduced;
    }
    return reduced.substring(0, reduced.length - 1);
  }, sanitizedText);

  return {
    transformedText: sanitizedText,
    targetText,
    solvedPrefix
  };
};

export const matchResult = ({ targetTexts, text, learnLanguageAlpha3 }) => {
  const { transformedText, targetText, solvedPrefix } = getClosestMatch({ text, targetTexts });
  return assesUserInput({ attemptText: transformedText, targetText, learnLanguageAlpha3 })
    .then((result) => {
      const solved = result !== RESULTS.INCORRECT;

      const selection = solved ? null :
        {
          start: solvedPrefix.length,
          end: transformedText.length
        };

      return {
        text: targetText,
        inputText: text,
        feedbackType: result,
        solved,
        selection
      };
    });
};

/**
 * omit leading spaces of words
 * @param {Array<{character: string, used: boolean, id: number}>} word
 */
export const trimWordStart = (word) => {
  const firstRealCharacterIdx = word.find(({ character }) => Boolean(character));
  if (firstRealCharacterIdx === -1) {
    return word;
  } else {
    return word.slice(firstRealCharacterIdx);
  }
};
