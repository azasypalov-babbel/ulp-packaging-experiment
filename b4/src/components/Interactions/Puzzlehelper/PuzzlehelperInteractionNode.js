import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Puzzlehelper from './Puzzlehelper';
import { interactiveNodeType } from '../shared/scene/scenePropTypes';
import { stripDiacritics, replaceSpecialCharacters } from '../../../lib/matchingUtils/normalise';
import { isInteractionKey } from '../../../lib/keyboardEvents';
import ControlledGap from '../shared/Write/ControlledGap';
import { INTERACTION_STATES } from '../shared/scene/useScene';
import PuzzlehelperFeedback from './PuzzlehelperFeedback';

const normaliseText = (string) => {
  const stripedChar = stripDiacritics(string).toLowerCase();
  return replaceSpecialCharacters(stripedChar);
};

export const getTokens = (node, order) => {
  if (!node) return [];

  const targetTexts = (node.targetChoices || [])
    .filter((choice) => choice.correct)
    .map((choice) => choice.sentence);

  const tokens = targetTexts[0]
    .split('')
    .map((character, idx) => ({
      idx,
      token: character,
      used: order.includes(idx)
    }));

  return tokens;
};

const PuzzlehelperInteractionNode = ({
  node,
  interactionState,
  onAttempt,
  learnLanguageAlpha3,
  onRetry,
  onProceed
}) => {
  const [order, setOrder] = useState([]);

  const tokens = getTokens(node, order);
  const latestText = order.map((id) => tokens[id].token).join('');

  const inputCommands =  {
    insert: (id) => {
      setOrder((order) => order.concat(id));
    },
    delete: () => {
      setOrder((order) => order.slice(0, order.length - 1));
    },
    deleteAll: () => {
      setOrder([]);
    },
    return: () => {
      onAttempt(latestText);
    }
  };

  const handleKeyDown = (event) => {
    if (isInteractionKey(event, ['Shift+Escape'])) {
      event.preventDefault();
      setOrder(tokens.map(({ idx }) => idx));
      onAttempt(tokens.map(({ token }) => token).join(''));
      return;
    }
    if (event.key === 'Backspace') {
      event.preventDefault();
      inputCommands.delete();
      return;
    }
    const index = tokens.findIndex(({ used, token }) =>
      !used &&
      normaliseText(token) === normaliseText(event.key)
    );

    if (index >= 0) {
      event.preventDefault();
      inputCommands.insert(index);
    }
  };

  if (!node.active) {
    return <ControlledGap
      {...node}
      targetChoices={[node.targetChoices[0]]}
      latestText={latestText}
    />;
  }

  return <>
    <ControlledGap
      {...node}
      targetChoices={[node.targetChoices[0]]}
      onKeyDown={handleKeyDown}
      active={interactionState !== INTERACTION_STATES.INACTIVE}
      showFeedback={interactionState === INTERACTION_STATES.FEEDBACK}
      latestText={latestText}
    />
    {interactionState === INTERACTION_STATES.INPUT && <Puzzlehelper
      tokens={tokens}
      inputCommands={inputCommands}
    />}

    {interactionState === INTERACTION_STATES.FEEDBACK &&
        <PuzzlehelperFeedback
          attempt={node?.attempt}
          latestText={latestText}
          tokens={tokens}
          inputCommands={inputCommands}
          learnLanguageAlpha3={learnLanguageAlpha3}
          onRetry={() => {
            inputCommands.deleteAll();
            onRetry();
          }}
          onProceed={onProceed}
        />}
  </>;
};

PuzzlehelperInteractionNode.propTypes = {
  node: interactiveNodeType.isRequired,
  interactionState: PropTypes.string.isRequired,
  onAttempt: PropTypes.func.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired,
  onRetry: PropTypes.func.isRequired,
  onProceed: PropTypes.func.isRequired
};


export default PuzzlehelperInteractionNode;
