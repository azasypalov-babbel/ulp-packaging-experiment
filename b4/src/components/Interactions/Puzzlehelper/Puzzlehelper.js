import React, { useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import shuffle from '../../../lib/shuffle';
import { splitBefore } from '../shared/Write/helpers';

import Button from '../../shared/Button/Button';
import PuzzlehelperSheet from './PuzzlehelperSheet';
import { StyledPuzzlehelper, StyledPuzzlehelperWrapper } from './styles';

const Puzzlehelper = ({
  tokens,
  inputCommands
}) => {
  const [seed] = useState(Math.random() * 10000);

  const {
    insert: onInsert,
    delete: onDelete,
    return: onReturn
  } = inputCommands;

  const shuffled = splitBefore(tokens, ({ token }) => token === ' ')
    .map(word => {
      const shuffleParams = { seed, notOriginalOrder: true }
      const startsWithSpace = word[0].token === ' ';
      return startsWithSpace ? [word[0], ...shuffle({ array: word.slice(1), ...shuffleParams })] : [...shuffle({ array: word, ...shuffleParams })];
    })
    .flat();

  return (
    <PuzzlehelperSheet
      onDelete={onDelete}
      onDone={onReturn}
    >
      <StyledPuzzlehelperWrapper>
        <StyledPuzzlehelper>
          {shuffled.map(({ token, used, idx }) => {
            return (
              <Button
                key={idx}
                onClick={() => onInsert(idx)}
                onMouseDown={e => e.preventDefault()} // Prevents focus change
                disabled={used}
                data-choice={token.toLowerCase()}
              >
                {token}
              </Button>
            );
          })}
        </StyledPuzzlehelper>
      </StyledPuzzlehelperWrapper>
    </PuzzlehelperSheet>
  );
};

Puzzlehelper.propTypes = {
  tokens: PropTypes.arrayOf(PropTypes.shape({
    token: PropTypes.string.isRequired,
    used: PropTypes.bool.isRequired,
    idx: PropTypes.number.isRequired
  })).isRequired,
  inputCommands: PropTypes.shape({
    insert: PropTypes.func.isRequired,
    delete: PropTypes.func.isRequired,
    deleteAll: PropTypes.func.isRequired,
    return: PropTypes.func.isRequired
  })
};

const mapStateToProps = ({ session }) => ({
  learnLanguageAlpha3: session.learnLanguageAlpha3
});

export default connect(mapStateToProps)(Puzzlehelper);
