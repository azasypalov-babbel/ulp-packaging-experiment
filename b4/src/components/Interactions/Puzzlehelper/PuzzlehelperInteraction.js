import React, { useEffect, useState } from 'react';
import useScene, { INTERACTION_STATES } from '../shared/scene/useScene';
import Text from '../../shared/Text';

import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import useHasLatestAttempt from '../shared/Write/hasLatestAttempt';
import { matchResult } from '../shared/Write/helpers';
import PuzzlehelperInteractionNode from './PuzzlehelperInteractionNode';
import { getActiveNode, getItem, getMistakeCount, isGap } from '../shared/scene/selectors';
import { play, requestPrompt } from '../shared/scene/actions';
import useFeedbackSounds from '../shared/useFeedbackSounds';

const quantiseStates = ({ active, activeNode, feedbackDismissed }) => {
  if (!active || !activeNode) return INTERACTION_STATES.INACTIVE;

  const { number, pending } = activeNode.attempt;
  if (number && !pending && !feedbackDismissed) return INTERACTION_STATES.FEEDBACK;

  return INTERACTION_STATES.INPUT;
};

export const PuzzlehelperInteraction = ({
  item,
  active,
  onFinish,
  learnLanguageAlpha3,
  onAttempt
}) => {
  const [
    scene,
    dispatch
  ] = useScene({
    learnLanguageAlpha3,
    item,
    matchResult
  });

  const [playFeedback, isFeedbackSoundDone] = useFeedbackSounds({
    resetDependencies: [getActiveNode(scene)]
  });
  const [feedbackDismissed, setFeedbackDismissed] = useState(true);

  const { nodes } = getItem(scene);
  const activeNode = getActiveNode(scene);

  const interactionState = quantiseStates({ active, activeNode, feedbackDismissed });

  const latestAttempt = useHasLatestAttempt(activeNode && activeNode.attempt);

  const mistakeCount = getMistakeCount(scene);

  const handleAttempt = (attemptText) => {
    if (feedbackDismissed === true) {
      setFeedbackDismissed(false);
      dispatch(requestPrompt(attemptText));
    }
  };

  const handleRetry = () => {
    setFeedbackDismissed(true);
  };

  const handleProceed = () => {
    setFeedbackDismissed(true);
    dispatch(play());
  };

  useEffect(() => {
    if (latestAttempt && active) {
      playFeedback(latestAttempt.solved);
      onAttempt({ attempt: latestAttempt });
    }
  }, [latestAttempt, onAttempt, active, playFeedback]);

  useEffect(() => {
    if (!activeNode) {
      dispatch(play());
    }
  }, [activeNode, dispatch]);

  useEffect(() => {
    if (scene.completed && active && isFeedbackSoundDone) {
      onFinish(mistakeCount);
    }
  }, [scene.completed, onFinish, mistakeCount, isFeedbackSoundDone, active]);

  return (
    <div>
      {nodes.map((node, idx) => isGap(node)
        ? <PuzzlehelperInteractionNode
          key={idx}
          node={node}
          learnLanguageAlpha3={learnLanguageAlpha3}
          interactionState={interactionState}
          onAttempt={handleAttempt}
          onRetry={handleRetry}
          onProceed={handleProceed}
        />
        : <Text key={idx} color="spaceGray" dangerouslySetInnerHTML={{ __html: node.text }} />
      )}
    </div>
  );
};

const mapStateToProps = ({ session }) => ({
  learnLanguageAlpha3: session.learnLanguageAlpha3
});

PuzzlehelperInteraction.propTypes = {
  item: PropTypes.object.isRequired,
  active: PropTypes.bool.isRequired,
  onFinish: PropTypes.func.isRequired,
  onAttempt: PropTypes.func.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired
};

PuzzlehelperInteraction.trainerConfig = {

};

export default connect(mapStateToProps)(PuzzlehelperInteraction);
