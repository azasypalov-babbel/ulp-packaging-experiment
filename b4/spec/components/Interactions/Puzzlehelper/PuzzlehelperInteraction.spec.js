import React from 'react';
import { shallow } from 'enzyme';

import { PuzzlehelperInteraction } from '../../../../src/components/Interactions/Puzzlehelper/PuzzlehelperInteraction';
import PuzzlehelperInteractionNode from
  '../../../../src/components/Interactions/Puzzlehelper/PuzzlehelperInteractionNode';

import useFeedbackSounds from '../../../../src/components/Interactions/shared/useFeedbackSounds';

import { matchResult } from '../../../../src/components/Interactions/shared/Write/helpers';
import { RESULTS } from '../../../../src/lib/matchingUtils/evaluate';

jest.mock('../../../../src/components/Interactions/Puzzlehelper/PuzzlehelperInteractionNode',
  () => 'PuzzlehelperInteractionNode');
jest.mock('../../../../src/components/Interactions/shared/Write/helpers');
jest.mock('../../../../src/components/Interactions/shared/useFeedbackSounds', () => jest.fn());

const mockPlayFeedback = jest.fn();
useFeedbackSounds.mockImplementation(() => [
  mockPlayFeedback,
  false
]);

const flushPromises = () => new Promise(setTimeout);

const defaultProps = {
  active: true,
  item: {
    type: 'phrase',
    learnLanguageText: '((dog)) der'
  },
  onFinish: jest.fn(),
  onAttempt: jest.fn(),
  learnLanguageAlpha3: 'DEU',
  shouldShowTransliteration: true
};

describe('Puzzlehelper Interaction', () => {
  describe('renders', () => {
    test('html', () => {
      const component = shallow(<PuzzlehelperInteraction {...defaultProps} />);
      expect(component).toMatchSnapshot();
    });

    test('a PuzzlehelperInteractionNode', () => {
      const component = shallow(<PuzzlehelperInteraction {...defaultProps} />);
      const gap = component.find('PuzzlehelperInteractionNode');
      expect(gap.exists()).toBe(true);
    });
  });

  describe('should pass props', () => {
    test('to PuzzlehelperInteractionNode', () => {
      const component = shallow(<PuzzlehelperInteraction {...defaultProps} />);
      const gap = component.find('PuzzlehelperInteractionNode');

      expect(gap.prop('node')).toBeDefined();
      expect(gap.prop('interactionState')).toBeDefined();
    });
  });

  describe('feedback sounds', () => {
    describe('with no mistake', () => {
      beforeEach(() => {
        matchResult.mockImplementationOnce(({ text }) =>
          Promise.resolve({
            text,
            inputText: text,
            feedbackType: RESULTS.CORRECT,
            solved: true
          })
        );
      });

      test('should play sound for correct attempt', async () => {
        const component = shallow(<PuzzlehelperInteraction {...defaultProps} />);
        const { onAttempt } = component.find(PuzzlehelperInteractionNode).props();

        onAttempt('dog');
        await flushPromises();

        expect(mockPlayFeedback).toHaveBeenCalledTimes(1);
        expect(mockPlayFeedback).toHaveBeenCalledWith(true);
      });
    });

    describe('when a mistake is made', () => {
      beforeEach(() => {
        matchResult.mockImplementationOnce(({ text }) =>
          Promise.resolve({
            text,
            inputText: text,
            feedbackType: RESULTS.INCORRECT,
            solved: false
          })
        );
      });

      test('should play sound for incorrect attempt', async () => {
        const component = shallow(<PuzzlehelperInteraction {...defaultProps} />);
        const { onAttempt } = component.find(PuzzlehelperInteractionNode).props();

        onAttempt('test');
        await flushPromises();

        expect(mockPlayFeedback).toHaveBeenCalledTimes(1);
        expect(mockPlayFeedback).toHaveBeenCalledWith(false);
      });
    });
  });

  describe('making attempts', () => {
    it('should track attempt', async () => {
      const component = shallow(<PuzzlehelperInteraction {...defaultProps} />);
      const { onAttempt } = component.find(PuzzlehelperInteractionNode).props();

      matchResult.mockImplementationOnce(({ text }) =>
        Promise.resolve({
          text: '((dog)) der',
          inputText: text,
          feedbackType: RESULTS.INCORRECT,
          solved: false
        })
      );
      onAttempt('user input');
      await flushPromises();

      expect(defaultProps.onAttempt).toHaveBeenCalledTimes(1);
      expect(defaultProps.onAttempt).toHaveBeenCalledWith(
        expect.objectContaining({
          attempt: {
            text: '((dog)) der',
            number: 1,
            solved: false,
            inputText: 'user input',
            feedbackType: RESULTS.INCORRECT,
            mistaken: true,
            pending: false,
            selection: undefined
          }
        })
      );
      defaultProps.onAttempt.mockClear();
    });
  });

  describe('on completion', () => {
    let component;
    beforeEach(async () => {
      component = shallow(<PuzzlehelperInteraction {...defaultProps} />);
      const { onAttempt } = component.find(PuzzlehelperInteractionNode).props();
      const { onProceed } = component.find(PuzzlehelperInteractionNode).props();

      matchResult.mockImplementationOnce(({ text }) =>
        Promise.resolve({
          text: '((dog)) der',
          inputText: text,
          feedbackType: RESULTS.CORRECT,
          solved: true
        })
      );

      onAttempt('der dog');
      await flushPromises();
      onProceed();
    });

    describe('before feedback sound is complete', () => {
      it('should not call onFinish', () => {
        expect(defaultProps.onFinish).not.toHaveBeenCalled();
      });
    });

    describe('after feedback sound is complete', () => {
      beforeEach(async () => {
        useFeedbackSounds.mockImplementation(() => [
          mockPlayFeedback,
          true
        ]);
        // Hack to update the component so it uses the above mocked value
        component.setProps({});
      });

      it('should call onFinish with mistakeCount', () => {
        const mistakeCount = 0;
        expect(defaultProps.onFinish).toHaveBeenCalledTimes(1);
        expect(defaultProps.onFinish).toHaveBeenCalledWith(mistakeCount);
      });
    });
  });
});
