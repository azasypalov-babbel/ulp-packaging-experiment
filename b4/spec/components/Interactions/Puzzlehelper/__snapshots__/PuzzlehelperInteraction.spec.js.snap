// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Puzzlehelper Interaction renders html 1`] = `
<div>
  <PuzzlehelperInteractionNode
    interactionState="INPUT"
    key="0"
    learnLanguageAlpha3="DEU"
    node={
      Object {
        "active": true,
        "attempt": Object {
          "feedbackType": null,
          "mistaken": false,
          "number": null,
          "pending": false,
          "selection": null,
          "solved": false,
          "text": null,
        },
        "isLastInteractiveNode": true,
        "targetChoices": Array [
          Object {
            "correct": true,
            "oddSolution": false,
            "sentence": "dog",
          },
        ],
        "text": null,
        "type": "gap",
      }
    }
    onAttempt={[Function]}
    onProceed={[Function]}
    onRetry={[Function]}
  />
  <styled.span
    color="spaceGray"
    dangerouslySetInnerHTML={
      Object {
        "__html": " der",
      }
    }
    key="1"
  />
</div>
`;
